Source/DafnyCore/Resolver.cs

```diff
@@ -9941,6 +9944,51 @@ public CheckDividedConstructorInit_Visitor(Resolver resolver)
       return true;
     }
 
+    bool CheckIfEqualityIsDefinitelyNotSupported(IndDatatypeDecl dt, Graph<IndDatatypeDecl/*!*/>/*!*/ dependencies) {
+
+      var scc = dependencies.GetSCC(dt);
+
+      foreach (var ctor in dt.Ctors) {
+
+        if (ctor.IsGhost) {
+          return true;
+        }
+
+        foreach (var arg in ctor.Formals) {
+          var anotherIndDt = arg.Type.AsIndDatatype;
+          if (arg.IsGhost ||
+              (anotherIndDt != null && anotherIndDt.EqualitySupport == IndDatatypeDecl.ES.Never) ||
+              arg.Type.IsCoDatatype ||
+              arg.Type.IsArrowType ||
+              (arg.Type.IsOpaqueType && !arg.Type.SupportsEquality)) {
+            return true;
+          }
+
+          if (anotherIndDt != null && !scc.Contains(anotherIndDt)) {
+            CheckIfEqualityIsDefinitelyNotSupported(anotherIndDt, dependencies);
+          }
+
+          foreach (var type in arg.Type.TypeArgs) {
```

Comment:
I think this check is too strict: we should only consider the cases where the type argument is needed for equality support

Reply:
That's an interesting point. I need to think more carefully about the different ways in which it could happen. In the example you give, where the type parameter is not used, I wonder: should we actually accept such a program in the first place? I guess that it doesn't hurt to have an unused type parameter, but isn't it always a typo?

Reply:
Refining a non-parametric type with a non-parametric one would be one non-typo example (but maybe not a very interesting one).

There's already code to track which type parameters are needed for equality, btw (see `NecessaryForEqualitySupportOfSurroundingInductiveDatatype`)

Reply:
Is there a type in your first sentence? :)

Reply:
There was, fixed!